CSC322-Grinnell/mayflower-directory
# Optimize directory by caching images

Resolves #143 

# Changes

- Keep avatar urls with the model and update them on a schedule before they expire
- Force update cached avatar urls when a profile is created or updated

# Testing

1. Run `rails db:migrate`
2. Run `rails test test/controllers/profiles_controller_test.rb test/jobs/cache_profile_avatars_job_test.rb`.  It should pass.  If you run just `rails test`, you will get two failures from the departments controller, but fixing those it outside of the scope of this PR
3. Restart the server.  On startup, it should spit out a bunch of database transactions about updating `cached_avatar_url`. Give it about 20 seconds to complete that job.
4. Go to the directory page.  It should load quickly (within about 3 seconds)
5. Create a new profile with an image.  The image should appear for the created profile
6. Edit a profile and change it's image.  The updated image should appear for that profile
7. Because it is on a 12 hour schedule, we can't really test that it refreshes on its own, so we should leave it in staging for at least 24 hours to be sure it works

## Fixing Issues with Migrations

If you already reviewed this PR, your database could get messed up because I changed one of the migrations.

If you have trouble running the server, run `rails db:migrate:redo VERSION=20190425193347` to re-execute migrations and add the correct database column.  This may delete any departments you have created.

If unit tests aren't running, try rebuilding the test database:
1. `rails db:drop RAILS_ENV=test`
2. `rails db:create RAILS_ENV=test`
3. `rails db:migrate RAILS_ENV=test`
